src/map/unit.cpp

```diff
@@ -1488,101 +1494,108 @@ int32 unit_warp(struct block_list *bl,int16 m,int16 x,int16 y,clr_type type)
  * Calculates the exact coordinates of a bl considering the walktimer
  * This is needed because we only update X/Y when finishing movement to the next cell
  * Officially, however, the coordinates update when crossing the border to the next cell
- * @param bl: Object to get the coordinates of
+ * The coordinates are stored in unit_data.pos together with the tick based on which they were calculated
+ * Access to these coordinates is only allowed through corresponding unit_data class functions
+ * This function makes sure calculation only happens when it's needed to save performance
  * @param tick: Tick based on which we calculate the coordinates
- * @param x: Will be set to the exact X value of the bl
- * @param y: Will be set to the exact Y value of the bl
- * @param sx: Will be set to the exact subcell X value of the bl
- * @param sy: Will be set to the exact subcell Y value of the bl
- * @return Whether the coordinates were set (true) or retrieving them failed (false)
  */
-bool unit_pos(block_list& bl, t_tick tick, int16 &x, int16 &y, uint8 &sx, uint8 &sy)
+void unit_data::update_pos(t_tick tick)
 {
-	unit_data* ud = unit_bl2ud(&bl);
+	// Check if coordinates are still up-to-date
+	if (DIFF_TICK(tick, this->pos.tick) < MIN_POS_INTERVAL)
+		return;
 
-	if (ud == nullptr)
-		return false;
+	if (this->bl == nullptr)
+		return;
 
-	if (ud->walkpath.path_pos >= ud->walkpath.path_len)
-		return false;
+	// Set initial coordinates
+	this->pos.x = this->bl->x;
+	this->pos.y = this->bl->y;
+	this->pos.sx = 8;
+	this->pos.sy = 8;
 
-	if (ud->walktimer == INVALID_TIMER)
-		return false;
+	// Remember time at which we did the last calculation
+	this->pos.tick = tick;
 
-	const TimerData* td = get_timer(ud->walktimer);
+	if (this->walkpath.path_pos >= this->walkpath.path_len)
+		return;
 
-	if (td == nullptr)
-		return false;
+	if (this->walktimer == INVALID_TIMER)
```

Comment:
Shoudnt we move this very important low-cost check almost to the top?

Reply:
If the walktimer is invalid, we still need to set the position correctly. The unit is just at the cell center currently. That's why it needs to come after the setting of the coordinates (and pos.tick).

Reply:
You mean because this->bl X/Y may be async to the new cached this->pos X/Y?

Reply:
Yes, we need to update the exact position also if the unit is not moving. Otherwise it might contain data from minutes ago. As we only calculate on demand.